tools/translation/tracker.js

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2021 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+/* eslint-disable no-underscore-dangle */
+/* global window, document, fetch, btoa */
+import { asyncForEach } from './utils.js';
+import { getLocales, getPathForLocale, getConfig } from './config.js';
+
+let config;
+
+async function init() {
+  if (config) return config;
+
+  const location = new URL(document.location.href);
+  const sp = location.searchParams.get('sp');
+  const owner = location.searchParams.get('owner');
+  const repo = location.searchParams.get('repo');
+  const ref = location.searchParams.get('ref');
+  if (sp && owner && repo && ref) {
+    const admin = (await getConfig()).admin;
+
+    const url = `${admin.api.preview.baseURI}/${owner}/${repo}/${ref}/?editUrl=${encodeURIComponent(sp)}`;
+    const resp = await fetch(url);
+    if (resp.ok) {
+      const json = await resp.json();
+      if (json && json.webPath) {
+        // compute the "real" filename, fallback to url last segment.
+        const str = json.source && json.source.sourceLocation ? json.source.sourceLocation : json.webPath;
```

Comment:
this can be written using optional chaining - `json?.source?.sourceLocation || json?.webPath`

Reply:
Yeah! I have learnt something today, thx @narcis-radu ;)